### PR TITLE
Fix OOM issue when input text is huge by splitting small chunks.

### PR DESCRIPTION
**Pull request by kenmasumitsu**
It is a PR for #131.

Without the changes, `hugeCharactersByDefaultMode` test fails with OOM.

### CODE REVIEW

**Review by mh-northlander — APPROVED** [review]

**Comment by kenmasumitsu** [comment]
I found, `offsetAtt` is not updated correctly in `incrementToken()`.  
I will fix it.

**Comment by kenmasumitsu** [comment]
> I found, `offsetAtt` is not updated correctly in `incrementToken()`. I will fix it.

I fixed it at 9d08c07410a8942e31899617ca0bd427115dead7.  
Could you please review it?

**Comment by mh-northlander** [comment]
It is still failing around the interaction with "analysis-icu" plugin.
I'm still investigating, but maybe we need something like [`readAsMuchAsCan`](https://github.com/WorksApplications/Sudachi/blob/bbf8ce708e8a195c3afa25b3ecd4ab986e55a621/src/main/java/com/worksap/nlp/sudachi/IOTools.java#L41) to fill buffer.

File: src/main/java/com/worksap/nlp/lucene/sudachi/ja/SudachiTokenizer.kt [hunk 1]
```diff
@@ -42,31 +44,57 @@ class SudachiTokenizer(
...
     clearAttributes()
-    val m = iterator.next() ?: return false
+
+    var m = iterator.next()
+    if (m == null) {
+      // To cope with huge text, it split into chunks (1 MB) for tokenize.
+      // TODO: Should split with meaningful delimitations instead of fixed size (1 MB).
+      val buffer = CharBuffer.allocate(1024)
+      // val buffer = CharBuffer.allocate(1 * 1024 * 1024)
+      val nread = input.read(buffer)
```
Comment (mh-northlander) [inline src/main/java/com/worksap/nlp/lucene/sudachi/ja/SudachiTokenizer.kt]: Let's use `com.worksap.nlp.sudachi.IOTools` here:
`val nread = IOTools.readAsMuchAsCan(input, buffer)`

Comment (kenmasumitsu) [reply src/main/java/com/worksap/nlp/lucene/sudachi/ja/SudachiTokenizer.kt]: I applied the changes.

**Comment by kenmasumitsu** [comment]
Integration test failed in another portion. 

```
Traceback (most recent call last):
  File "/home/runner/work/elasticsearch-sudachi/elasticsearch-sudachi/test-scripts/30-test-docs.py", line 74, in <module>
    Test(parse_args()).run()
  File "/home/runner/work/elasticsearch-sudachi/elasticsearch-sudachi/test-scripts/30-test-docs.py", line 30, in run
    obj()
  File "/home/runner/work/elasticsearch-sudachi/elasticsearch-sudachi/test-scripts/30-test-docs.py", line 38, in test107Daigaku
    self.assertEq(107, docs["hits"]["total"]["value"])
  File "/home/runner/work/elasticsearch-sudachi/elasticsearch-sudachi/test-scripts/30-test-docs.py", line 23, in assertEq
    raise Exception("a != b", a, b)
Exception: ('a != b', 107, 104)
Error: Process completed with exit code 1.
```
https://github.com/WorksApplications/elasticsearch-sudachi/actions/runs/9492521212/job/26166010901?pr=132#step:11:428

In my investigation, it seems test code issue.  (I believe it is not related to this PR.)

`test-scripts/20-put-docs.py` calls PUT requests 16051 times to register documents. But some of requests failed due to status code `429` (Too Many Requests). 
As a result, `test-scripts/30-test-docs.py` failed.

I modified `test-scripts/20-put-docs.py` to call POST request once with `bulk` API.  Then the test passed.

**Comment by mh-northlander** [comment]
Ignore test es:8.13.4, it is failing because of the secret token issue.

The previous test (20-put-docs.py) seems failed due to the performance issue. Allocating 1M-charbuffer each time costs a lot memory and slows down indexing.
Although the new test shows the proper use of API reduces the problem, I also want to reduce the memory usage of char buffer itself.

Could you let the tokenizer not to allocate too much memory, e.g. starting with small buffer size and grow it when necessary at the maximum of 1M?

**Comment by kenmasumitsu** [comment]
Still there is one failure at https://github.com/WorksApplications/elasticsearch-sudachi/actions/runs/9495843610/job/26169450690?pr=132#step:8:1158

The cause of this error is that the `SONAR_TOKEN` in `secrets` is not accessible.
The workflow from the forked repository does not have access to the `secrets`, so this error cannot be fixed for now.
https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks

**Comment by kenmasumitsu** [comment]
> e.g. starting with small buffer size and grow it when necessary at the maximum of 1M?

I think it is not good idea, because it becomes complex. 

For the performance, I suggest 'val buffer = CharBuffer.allocate(1 * 1024 * 1024)' move to member variable from local variable.
It allocates memory once on 'new SudachiTokenizer()'.

**Comment by mh-northlander** [comment]
Moving char buffer to the tokenizer's member variable seems working in my local env.
Could you revert the update of `20-put-docs` with this change?
I will accept it if the test also passes in the github actions env.

**Comment by kenmasumitsu** [comment]
> Moving char buffer to the tokenizer's member variable seems working in my local env.

I will do. But it is effective on performance, only if `incrementToken()` called multiple times.

> starting with small buffer size and grow it when necessary at the maximum of 1M?

I will consider this approach.

> Could you revert the update of 20-put-docs with this change?

Yes

**Comment by kenmasumitsu** [comment]
> > Moving char buffer to the tokenizer's member variable seems working in my local env.
> 
> I will do. But it is effective on performance, only if `incrementToken()` called multiple times.
> 
> > starting with small buffer size and grow it when necessary at the maximum of 1M?
> 
> I will consider this approach.

I implemented them at 36cedbf6b8f8a7a78dd03655d18612f4e7519a7f.

**Comment by mh-northlander** [comment]
@kenmasumitsu
LGTM, thank you very much.

**Pull request merged**